siriuspy/siriuspy/devices/fpmosc.py

```diff
@@ -0,0 +1,124 @@
+"""."""
+
+import numpy as _np
+
+from ..currinfo.csdev import Const as _Const
+from .device import Device as _Device
+
+
+class FPMOsc(_Device):
+    """."""
+
+    OffOn = _Const.OffOn
+
+    class DEVICES:
+        """Devices names."""
+
+        SI = 'SI-Glob:DI-FPMOsc'
+        ALL = (SI, )
+
+    PROPERTIES_DEFAULT = (
+        'UpdateTime-SP',
+        'UpdateTime-RB',
+        'FiducialOffset-SP',
+        'FiducialOffset-RB',
+        'FillPatternRef-SP',
+        'FillPatternRef-RB',
+        'UniFillEqCurrent-Mon',
+        'ErrorRelStd-Mon',
+        'ErrorKLDiv-Mon',
+        'FillPatternRef-Mon',
+        'FillPattern-Mon',
+        # 'TimeOffset-Mon',
+        # 'Time-Mon',
+        # 'Raw-Mon',
+        # 'RawAmp-Mon',
+        # 'RawTime-Mon',
+    )
+
+    def __init__(self, devname=None, props2init='all'):
+        """."""
+        # check if device exists
+        devname = devname or FPMOsc.DEVICES.SI
+        if devname not in FPMOsc.DEVICES.ALL:
+            raise NotImplementedError(devname)
+        super().__init__(devname, props2init=props2init)
+
+    @property
+    def update_time(self):
+        """."""
+        return self['UpdateTime-RB']
+
+    @update_time.setter
+    def update_time(self, value):
+        """."""
+        self['UpdateTime-SP'] = value
+
+    @property
+    def fiducial_offset(self):
+        """."""
+        return self['FiducialOffset-RB']
+
+    @fiducial_offset.setter
+    def fiducial_offset(self, value):
+        self['FiducialOffset-SP'] = value
+
+    @property
+    def fill_pattern_ref(self):
+        """."""
+        return self['FillPatternRef-RB']
+
+    @fill_pattern_ref.setter
+    def fill_pattern_ref(self, value):
+        """."""
+        if not isinstance(value, (_np.ndarray, list, tuple)):
+            return
+        value = _np.array(value)
+        if value.size != _Const.FP_HARM_NUM:
+            return
+        self['FillPatternRef-SP'] = value
```

Comment:
Shouldn't this raise some warning for the user that the reference was not set?

Reply:
While our logging system is not standardized, I prefer not to print stuff. But maybe we could raise an exception...

Reply:
done!